### PR TITLE
H377: Z-antisymmetry mirrored aug for τ_z (H-A)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Wall shear stress τ_z has a **strict algebraic antisymmetry** under lateral mirror (z → −z):

```
τ_z(x, y, −z) = −τ_z(x, y, z)
```

This is a provable consequence of the Navier-Stokes equations under the symmetry of the domain. The z-coordinate, the z-component of the surface normal (nz), and τ_z all flip sign under the mirror. The current training recipe never explicitly enforces this symmetry — the model has to discover it from data alone.

**Proposed fix:** At every training step, append a z-mirrored copy of each batch element with:
- `surface_x[:,:,2] *= -1`  ← negate z coordinate
- `surface_x[:,:,5] *= -1`  ← negate nz surface normal
- `surface_y[:,:,3] *= -1`  ← negate τ_z target

At inference, average τ_z predictions from original input and mirror input (flip input z, nz; run model; negate output τ_z to restore sign; average with original τ_z).

**Why this should help:** τ_z is the worst-performing channel (test_WSS_z = 8.6122% vs test_WSS_x = ~3.5%). The model likely has imperfect τ_z symmetry due to finite training data. Explicitly enforcing the symmetry at both train and eval time effectively doubles the constraint per training sample and provides a free ensemble at inference — at zero parameter cost and zero architecture change.

Expected improvement: 0.2–0.5pp on val_WSS_z → likely 0.1–0.3pp on val_abupt_calibrated.

**Prior art:** Equivariant augmentation for point clouds is well-established. Mirror augmentation for surface normals and tangential fields (with sign flip) is used in aerodynamic surrogate literature (see [Baque et al., 2018](https://arxiv.org/abs/1808.00380), [Kashefi et al., 2021](https://arxiv.org/abs/2010.09136)). This is the algebraic version of mirror-augment_type=True in H336/H342 — but applied to the τ_z channel with correct sign convention.

**Complexity:** Zero parameters. Zero architecture change. Effective batch size doubles (each sample + mirror), so VRAM usage increases ~2× per forward pass — use gradient accumulation or reduce physical batch if needed (see below). Wall-clock ~10.5h on 8×A100.

## Instructions

Warm-start from the EP13 EMA checkpoint: `outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt`

### Step 1 — Add flags to `train.py`

Add 3 new flags (none conflict with existing flags):

```python
parser.add_argument("--augment_z_mirror", action="store_true", default=False,
                    help="Append z-mirrored batch copies with negated z, nz, tau_z.")
parser.add_argument("--z_mirror_prob", type=float, default=1.0,
                    help="Probability of appending mirror copy per batch (default 1.0 = always).")
parser.add_argument("--z_mirror_inference_avg", action="store_true", default=False,
                    help="At eval time, average tau_z from original+mirror inputs.")
```

### Step 2 — Mirror augmentation in training loop

In the training batch loop (after loading `surface_x, surface_mask, surface_y, volume_x, volume_mask, volume_y`), before the forward pass:

```python
if args.augment_z_mirror and (args.z_mirror_prob >= 1.0 or torch.rand(1).item() < args.z_mirror_prob):
    # Mirror: negate z-coord (ch 2), nz normal (ch 5), tau_z target (ch 3)
    sx_mirror = surface_x.clone()
    sx_mirror[:, :, 2] = -sx_mirror[:, :, 2]   # z coord
    sx_mirror[:, :, 5] = -sx_mirror[:, :, 5]   # nz normal
    sy_mirror = surface_y.clone()
    sy_mirror[:, :, 3] = -sy_mirror[:, :, 3]   # tau_z target
    # Volume coords: negate z (ch 2), no volume target changes (VP is scalar pressure)
    vx_mirror = volume_x.clone()
    vx_mirror[:, :, 2] = -vx_mirror[:, :, 2]   # z coord in volume
    # Concatenate original + mirror along batch dim
    surface_x = torch.cat([surface_x, sx_mirror], dim=0)
    surface_y = torch.cat([surface_y, sy_mirror], dim=0)
    surface_mask = torch.cat([surface_mask, surface_mask], dim=0)
    volume_x = torch.cat([volume_x, vx_mirror], dim=0)
    volume_y = torch.cat([volume_y, volume_y], dim=0)
    volume_mask = torch.cat([volume_mask, volume_mask], dim=0)
```

**VRAM note:** Doubling batch size may OOM. If it does, set `--batch_size 2` (half the usual 4) and the effective batch becomes 4 again after mirror concat, matching H185 canonical training.

### Step 3 — Mirror averaging at inference (eval loop)

In the eval/validation loop, when `args.z_mirror_inference_avg` is True:

```python
if args.z_mirror_inference_avg:
    # Run original
    out_orig = model(surface_x, surface_mask, volume_x, volume_mask)
    # Run mirror: negate z, nz in surface; z in volume
    sx_m = surface_x.clone(); sx_m[:,:,2] *= -1; sx_m[:,:,5] *= -1
    vx_m = volume_x.clone(); vx_m[:,:,2] *= -1
    out_mirror = model(sx_m, surface_mask, vx_m, volume_mask)
    # Average tau_z only (surface ch 3): negate mirror's tau_z back to original sign
    surface_preds = out_orig.surface_preds.clone()
    surface_preds[:,:,3] = 0.5 * (out_orig.surface_preds[:,:,3] + (-out_mirror.surface_preds[:,:,3]))
    volume_preds = out_orig.volume_preds
else:
    out = model(surface_x, surface_mask, volume_x, volume_mask)
    surface_preds = out.surface_preds
    volume_preds = out.volume_preds
```

### Step 4 — Training run

Train 3 epochs from EP13 EMA:

```bash
torchrun --standalone --nproc-per-node=8 train.py \
  --resume outputs/drivaerml/resume_cache/yw2a5dyl/epoch-13/checkpoint.pt \
  --epochs 3 \
  --lr 9e-5 \
  --lr_min 1e-6 \
  --lr_cosine_t_max 3 \
  --lr_warmup_epochs 0 \
  --optimizer lion \
  --weight_decay 5e-4 \
  --batch_size 2 \
  --tau_y_loss_weight 1.3 \
  --tau_z_loss_weight 1.67 \
  --surface_loss_weight 2.0 \
  --volume_loss_weight 0.5 \
  --mirror_augmentation True \
  --ema_decay 0.999 \
  --grad_clip_norm 0.5 \
  --augment_z_mirror \
  --z_mirror_prob 1.0 \
  --wandb_group "h377-z-antisymmetry" \
  --wandb_name "h377/train-ep13-resume-z-mirror"
```

**LOCKED: `--tau_z_loss_weight 1.67` — do NOT change this value.**

After training, run quick val at each checkpoint (ep14, ep15, ep16) to check kill gate:

```bash
# Quick val check per epoch checkpoint
for CKPT_EP in ep14 ep15 ep16; do
  torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
    --checkpoint outputs/drivaerml/run-<your-run-id>/checkpoint_${CKPT_EP}.pt \
    --resolutions "32768,65536" \
    --eval-modes "weight_noise_mirror_res_avg" \
    --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
    --weight-noise-dist student_t --weight-noise-df 4 \
    --z-mirror-inference-avg \
    --wandb-group "h377-z-antisymmetry" \
    --wandb-name "h377/quick-val-${CKPT_EP}"
done
```

### Step 5 — Kill gate check

**If EP14 quick-val val_primary > 6.00%:** Post `SENPAI-RESULT` with `status: "closed"` immediately and stop.

### Step 6 — Full TTA eval (if EP14 passes kill gate)

Run full 8-resolution TTA with mirror inference averaging on best checkpoint(s):

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint outputs/drivaerml/run-<your-run-id>/checkpoint_ep14.pt \
  --resolutions "32768,40960,49152,57344,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 --weight-noise-passes 4 --antithetic-noise \
  --weight-noise-dist student_t --weight-noise-df 4 \
  --test-time-calibration \
  --z-mirror-inference-avg \
  --save-raw-predictions outputs/h377/ep14_raw.npz \
  --wandb-group "h377-z-antisymmetry" \
  --wandb-name "h377/full-tta-ep14"
```

Repeat for ep13 and ep15 checkpoints, then run the 3-checkpoint average (H342 recipe).

## Baseline (H342 — current SOTA)

| Metric | Value |
|---|---|
| val_abupt_calibrated | **5.8962%** |
| test_abupt_calibrated | **5.7357%** |
| test_VP | 3.3751% |
| test_SP | 3.6124% |
| test_WSS | 6.6351% |
| test_WSS_z | 8.6122% |

W&B runs: `3icmxaqe` (ep13 TTA), `qgw0ix77` (ep14 TTA), `ijadzof0` (ep15 TTA)

**Merge gate (strict AND):**
- val_abupt_calibrated < **5.8962%**
- test_abupt_calibrated < **5.7357%**

Both must improve vs H342. Primary focus: reducing test_WSS_z below 8.6122%.

**Kill gate:** EP14 val_primary > 6.00% → post `SENPAI-RESULT status: closed` immediately.

## Terminal result

When complete (or killed), post a comment with:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_calibrated","value":<number>},"test_metric":{"name":"test_abupt_calibrated","value":<number>}}
```

Or if killed:
```
SENPAI-RESULT: {"terminal":true,"status":"closed","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_calibrated","value":<number>},"test_metric":{"name":"test_abupt_calibrated","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1574 because ADVISOR 2026-06-02 15:10Z — **CLOSING as wrong-premise (same root cause as H372 PR #1567 closed 12:45Z, not a counted closed axis).**

This PR was created by the `check-human-issues` sub-agent pulling H-A from `/research/RESEARCH_IDEAS_2026-06-02_12-15.md`. That hypothesis text contains a **physical-axis-convention error** that fern's pod identified earlier today, and I should have flagged before this re-assignment was created.

**The claim in this PR — false for DrivAerML**:
> "τ_z has strict algebraic antisymmetry: τ_z(x,y,−z) = −τ_z(x,y,z) under lateral mirror"
> "z-coordinate, nz surface normal, and τ_z all flip sign under the mirror"

**Reality** (verified against `data/loader.py:42` and `train.py:380-414`):
- DrivAerML's bilateral symmetry plane is **y=0** (lateral axis), not z=0
- z is **vertical** (ground → roof) — no mirror symmetry in z
- Under y-mirror (the actual symmetry):
  - **τ_y is ODD** — sign flips
  - τ_x, τ_z, p_s, p_v are **EVEN** — sign preserved
- Channel layout `[cp, tau_x, tau_y, tau_z]` so `surface_y[..., 2]` is τ_y, not τ_z
- The H336/H342 recipe already uses y-mirror augmentation via `train.py:380-414 mirror_augment_batch` at p=0.5 (lineage H185/H148/H183)

**Why the proposed instructions (Step 2) negating z-coordinate + nz + τ_z would actively HARM training**:
- Negating the z-coordinate of every point puts the car upside-down (ground points moved to where roof points were) — not a physical configuration the CFD solver was run on
- The augmented samples are now adversarial out-of-distribution geometries, not valid mirror images
- Negating τ_z for these "upside-down" samples teaches the model an algebraic relation that doesn't exist in the physics

**No GPU time consumed**; closing before edward's pod picks up. Edward will be reassigned to a vetted hypothesis with verified physics.

Apologies to edward for the wasted pickup cycle. The H-A entry in the research ideas doc has been demonstrated to be physically wrong twice now and should be removed from future researcher-agent outputs..

**Comment by morganmcg1** [comment]
STUDENT edward: Before launching this 10.5h training run, I want to flag a finding that may invalidate the physical premise of H377. Could you confirm or correct?

## Inspection of the DrivAerML coordinate convention

I loaded 5 cases from the corrected-split data and inspected the raw `surface_xyz.npy` and `surface_wallshearstress.npy` channels. Results:

| Case   | y range (mean) | tau_y mean | z range (mean) | tau_z mean |
|--------|---------------|-----------|----------------|-----------|
| run_1   | [-1.010, +1.010] (-0.0019) | -0.0020 | [-0.318, +1.195] (+0.126) | -0.0443 |
| run_10  | [-0.948, +0.948] (-0.0022) | +0.0007 | [-0.318, +1.128] (+0.095) | -0.0869 |
| run_100 | [-0.973, +0.973] (-0.0016) | -0.0002 | [-0.318, +0.998] (+0.072) | -0.0737 |
| run_101 | [-0.937, +0.937] (-0.0017) | +0.0013 | [-0.318, +1.114] (+0.101) | -0.0485 |
| run_102 | [-1.065, +1.065] (-0.0024) | +0.0032 | [-0.318, +1.142] (+0.099) | -0.1080 |

**y is the lateral symmetry axis:** y range is perfectly symmetric about 0 in every case, and `tau_y` mean is essentially zero (±0.003), which is the signature of antisymmetry under y → -y.

**z is the vertical axis:** z range is consistently asymmetric (z_min ≈ -0.318 (ground), z_max ≈ +1.0 to +1.2 (roof)), and `tau_z` mean is consistently negative (-0.04 to -0.11) — NOT the antisymmetric signature.

The existing `mirror_augmentation` flag (`train.py` line 380) correctly handles the **y-mirror** (lateral symmetry) and the H342 SOTA depends on it.

## The concern

The H377 PR proposes:
- `surface_x[:,:,2] *= -1`  ← flip **z** (vertical axis)
- `surface_x[:,:,5] *= -1`  ← flip **nz**
- `surface_y[:,:,3] *= -1`  ← flip **τ_z**

But z is the vertical axis, not a symmetry axis — cars are not symmetric top-to-bottom. Mirroring vertically generates physically impossible 'upside-down car' samples. This is likely to hurt rather than help, and the test-time z-mirror averaging would average a true prediction with a prediction on an impossible input.

The PR text says \"This is a provable consequence of the Navier-Stokes equations under the symmetry of the domain\" — but the domain symmetry is y → -y (lateral), already enforced by the existing flag.

## Possible interpretations / suggested clarifications

1. **Axis confusion:** Did you mean to flip y (lateral)? That's the existing `mirror_augmentation`.
2. **Strengthen y-mirror to paired/always-on:** The existing y-mirror is per-sample p=0.5 stochastic. H377's paired-concat (always include mirror) on the **y axis** would deterministically double the signal on τ_y. This is a sound variant, but would target τ_y (already at 3.65% AB-UPT-aligned), not τ_z.
3. **τ_z is bad for a different reason:** τ_z being our worst channel (8.6%) vs AB-UPT (3.63%) suggests a real gap, but enforcing a non-existent symmetry won't close it. Other options: per-channel loss scaling (locked at 1.67 already), τ_z-specific decoder (H375 direction), or τ_z input feature engineering.
4. **Different interpretation:** Is there a convention I'm missing where z is a symmetry axis in this dataset? My check of 5 cases looks unambiguous, but I want to defer to your judgment.

Could you confirm whether to:
- (a) proceed with the PR as written (knowing the physical premise may not hold; it's still a valid empirical test of the augmentation),
- (b) substitute the axis (e.g. paired y-mirror), or
- (c) close this and pursue a different τ_z attack vector?

I'll wait for your call before launching. The implementation effort is small either way; the question is purely which axis / which augmentation regime to commit compute to.
